jmxtrans-output/jmxtrans-output-kafka/src/main/java/com/googlecode/jmxtrans/model/output/kafka/KafkaWriter.java

```diff
@@ -117,14 +119,17 @@ protected void internalWrite(Server server, Query query, ImmutableList<Result> r
 					String message = createJsonMessage(server, query, typeNames, result, values, value);
 					for(String topic : this.topics) {
 						log.debug("Topic: [{}] ; Kafka Message: [{}]", topic, message);
-						producer.send(new KeyedMessage<String, String>(topic, message));
-						producer.close();
+						messages.add(new KeyedMessage<String, String>(topic, message));
 					}
 				} else {
 					log.warn("Unable to submit non-numeric value to Kafka: [{}] from result [{}]", value, result);
 				}
 			}
 		}
+
+		// Batch them all
+		producer.send(messages);
+		producer.close();
```

Comment:
Looking at the code in `kafka.javaapi.producer.Producer` it looks like if the producer is shutdown, it does not restart. Since we reuse the same producer for the lifetime of our `KafkaWriter`, this does not seem like a good idea. The `Producer` should probably be closed only when the `KafkaWriter` is closed (by overriding the `close()` method). It also looks like the `Producer` class is deprecated and we should replace it with `KafkaProducer`.

That being said, I have no experience with Kafka, so I might be completely wrong...

Sorry for the delay in my review!